pandas/core/generic.py

```diff
@@ -1948,6 +1948,129 @@ def tail(self, n=5):
+            Missing values in the weights column will be treated as zero. 
+            inf and -inf values not allowed. 
+        random_state : int or numpy.random.RandomState, optional
+            Seed for the random number generator (if int), or numpy RandomState 
+            object.
+        axis : int, optional
+            Axis to sample. Default = 0.
+            
+        Returns
+        -------
+        Same type as caller. 
+        """
+        
+        ###        
+        # Processing axis argument
+        ###
+        
+        # This checks validity of axis argument.
+        if axis > (self.ndim - 1):
+            raise ValueError("Axis argument out of range")
+    
+        # Store length of relevant axis of object. 
+        axis_length = self.shape[axis]    
+    
+        ###
+        # Processing random_state argument
+        ###
+    
+        # random_state may be either a RandomState object, 
+        # integer, or None. Otherwise raise an error. 
+     
+        if isinstance(random_state, int):
+            rs = np.random.RandomState(random_state)
+        elif isinstance(random_state, np.random.RandomState):
+            rs = random_state
+        elif random_state is None:
+            rs = np.random.RandomState()
+        else:
+            raise ValueError("Seed must be either an integer or numpy RandomState")
+    
+        ###
+        #  Process weight argument
+        ###
+    
+        # Check weights for compliance
+        if weights is not None:
+                                      
+            # Strings acceptable if not a series
+            if isinstance(weights, string_types): 
+
+                if self.ndim > 1 :
+                    try:
+                        weights = self[weights]
+                    except KeyError:
+                        raise KeyError("String passed to weights not a valid column name")
+
+                else:
+                    raise ValueError("Strings cannot be passed as weights when sampling from a Series.")
+
+            #normalize format of weights to ndarray. 
+            weights = np.asarray(weights, dtype = np.float)
+                        
+            # Check length (numpy does this, but has confusing errors with different argument labels.)
+            if len(weights) != axis_length:
+                raise ValueError("Weights and axis to be sampled must be of same length")
+
+            # No infs allowed. The np.nan_to_num() command below would make these large values
+            # which is pretty unintuitive. 
+            if np.isinf(weights).any():
+                raise ValueError("weight vector many not include `inf` values")
+                
+            if (weights < 0).any():
+                raise ValueError("weight vector many not include negative values")
+
+            # If has nan, set to zero. Already know there are no infs. 
+            weights = np.nan_to_num(weights)            
```

Comment:
As above, just wanted to avoid moving into a Series, then back to an array for np.choice().